src/main/java/ch/njol/skript/events/EvtGrow.java

```diff
@@ -18,85 +18,216 @@
+		eventRestriction = parseResult.mark; // ANY, STRUCTURE or BLOCK
+		actionRestriction = matchedPattern; // OF, FROM, INTO, or FROM_INTO
+
+		switch (actionRestriction) {
+			case OF:
+				if (eventRestriction == STRUCTURE) {
+					fromTypes = (Literal<Object>) args[0];
+				} else if (eventRestriction == BLOCK) {
+					fromTypes = (Literal<Object>) args[1];
+				}
+				break;
+			case FROM:
+				fromTypes = (Literal<Object>) args[0];
+				break;
+			case INTO:
+				if (eventRestriction == STRUCTURE) {
+					toTypes = (Literal<Object>) args[0];
+				} else if (eventRestriction == BLOCK) {
+					toTypes = (Literal<Object>) args[1];
+				}
+				break;
+			case FROM_INTO:
+				fromTypes = (Literal<Object>) args[0];
+				if (eventRestriction == STRUCTURE) {
+					toTypes = (Literal<Object>) args[1];
+				} else if (eventRestriction == BLOCK) {
+					toTypes = (Literal<Object>) args[2];
+				}
+				break;
+			default:
+				assert false;
+				return false;
+		}
 		return true;
 	}
 	
 	@Override
-	public boolean check(final Event e) {
-		if (evtType == STRUCTURE  && types != null && e instanceof StructureGrowEvent) {
-			return types.check(e, new Checker<StructureType>() {
-				@Override
-				public boolean check(final StructureType t) {
-					return t.is(((StructureGrowEvent) e).getSpecies());
+	public boolean check(Event event) {
+		// Exit early if we need fromTypes, but don't have it
+		if (fromTypes == null && actionRestriction != INTO)
+			// We want true for "on grow:", false for anything else
+			// So check against "OF", which is the first pattern; the one that allows "on grow:"
+			return actionRestriction == OF;
+
+		// Can exit early if we're checking against a structure, but the event isn't a structure grow event
+		// Can also exit early if we're checking against a block, but the event isn't a block grow event AND we're not checking for OF
+		// With OF, we can have `on grow of sapling` or `big mushroom` be a StructureGrowEvent
+		if (eventRestriction == STRUCTURE && !(event instanceof StructureGrowEvent)) {
+			return false;
+		} else if (eventRestriction == BLOCK && !(event instanceof BlockGrowEvent) && actionRestriction != OF) {
+			return false;
+		}
+
+		switch (actionRestriction) {
+			case OF:
+				return checkFrom(event, fromTypes) || checkTo(event, fromTypes);
+			case FROM:
+				return checkFrom(event, fromTypes);
+			case INTO:
+				return checkTo(event, toTypes);
+			case FROM_INTO:
+				return checkFrom(event, fromTypes) && checkTo(event, toTypes);
+			default:
+				assert false;
+				return false;
+		}
+	}
+
+	private static boolean checkFrom(Event event, Literal<Object> types) {
+		// treat and lists as or lists
+		if (types instanceof LiteralList)
+			((LiteralList<Object>) types).setAnd(false);
+		if (event instanceof StructureGrowEvent) {
+			Material sapling = ItemUtils.getTreeSapling(((StructureGrowEvent) event).getSpecies());
+			Skript.info("sapling: " + sapling);
```

Comment:
Oops! Why are we logging this?
If this is supposed to be some kind of debug output there's a special `debug()` method, otherwise I think this will just log always.